### PR TITLE
Use Karma and webpack to test gringotts

### PR DESCRIPTION
**Pull request by bzalasky**
- Remove Grunt, RequireJS, and related dependencies
  - Use ES modules
  - Configure Karma and webpack
  - TODO: Add a release script and rebase to include 
     Eugene's recent changes

### CODE REVIEW

File: npm-debug.log.1386857604 [hunk 1]
```diff
@@ -0,0 +1,24 @@
+0 info it worked if it ends with ok
```
Comment (eugenet8k) [inline npm-debug.log.1386857604]: we should exclude this file

Comment (bzalasky) [inline npm-debug.log.1386857604]: Yeah, I'll add a note to do that, along with the linters.

**Comment by eugenet8k** [comment]
@bzalasky can we configure karma to use "dot" reporter in "development" mode, but the full reporter in "production" (travis) mode?

Also, can we rename those modes to just "dev" and "prod"? =)

**Comment by eugenet8k** [comment]
I would suggest creating a folder `templates/progress-dialog/` and put everything in there having partials on the same dir level as views.

File: src/views/dialog-view.coffee [hunk 2]
```diff
@@ -1,23 +1,22 @@
...
-      'title', 'text', 'buttons'
-    ]
-    events:
-      'click button': (e) ->
-        $el = $(e.currentTarget)
-        @buttons.forEach (b) =>
-          b.click.call this, e if b.click and $el.hasClass b.className
+export default class DialogView extends ModalView
+  className: 'dialog'
+  template: require 'dialog.hbs'
```
Comment (eugenet8k) [inline src/views/dialog-view.coffee]: Is this a good practice to mix `include` and `import`. I wonder if we should import template in the top of every module like:

```
import template from 'dialog.hbs`

export default class DialogView extends ModalView
    template: template
```

But current solution looks more compact, I kinda like it better.

Comment (bzalasky) [inline src/views/dialog-view.coffee]: ES imports won't work with non Coffee/JS files. We'd have to do `template = require 'dialog.hbs'`. If the template is only being used in one place, it's fine to do it like this. If we have a reason to assign it, we can do that as well, as we do when requiring the sorting and pagination partials.

**Comment by bzalasky** [comment]
@eugenet8k Regarding dev and prod, `NODE_ENV` will need to be `production` for React to switch to it's optimized build in prod.

**Comment by bzalasky** [comment]
That doesn't matter so much here, but I'm just in the habit I suppose.

**Comment by eugenet8k** [comment]
@bzalasky if there is expectations on "production" to have as full word, I don't mind using "development" and "production" as is then.

**Comment by eugenet8k** [comment]
@bzalasky what version of NodeJS we need to use to rely on `package-log.json`? If it's higher than 6.9.2 which is set in `.nvmrc` then we need to update the version in that file as well.

**Comment by bzalasky** [comment]
@eugenet8k i'm using 7.4.0, but it should work with LTS.

**Comment by eugenet8k** [comment]
@bzalasky so, do you think we should update version in `.nvmrc`?

File: package.json [hunk 3]
```diff
@@ -20,52 +21,51 @@
     "node": ">=5.0"
   },
   "scripts": {
-    "start": "grunt",
-    "test": "grunt test-ci"
+    "start": "bin/test.sh",
```
Comment (eugenet8k) [inline package.json]: Is this a convention to have shell scripts in `/bin` folder? I wonder if we can use `/scripts` for that.

File: src/index.coffee [hunk 4]
```diff
@@ -0,0 +1,54 @@
+require 'lib/view-helper'
```
Comment (eugenet8k) [inline src/index.coffee]: What's the use for this line?

Comment (eugenet8k) [inline src/index.coffee]: Oh I see it just register those helpers on Handlebars object.

File: src/lib/mixin-builder.coffee [hunk 5]
```diff
@@ -1,20 +1,19 @@
-define (require) ->
-  mixinHelper = require './mixin-helper'
+mixinHelper = require './mixin-helper'
 
+###*
+  * A helper class that gets a list of mixins and creates
+  * a chain of inheritance.
+###
+module.exports = class MixinBuilder
```
Comment (eugenet8k) [inline src/lib/mixin-builder.coffee]: Is this a preferred  syntax for defining exports for classes? I wonder if we could do:

```
class MixinBuilder
   ...

module.exports = MixinBuilder
```

File: src/lib/view-helper.coffee [hunk 6]
```diff
@@ -1,128 +1,127 @@
-define (require) ->
-  handlebars = require 'handlebars'
-  moment = require 'moment'
-  utils = require './utils'
-  $ = require 'jquery'
+import handlebars from 'handlebars'
```
Comment (eugenet8k) [inline src/lib/view-helper.coffee]: What's the reason of using `import` here but `require` in other files?

File: src/templates/progress-dialog.hbs [hunk 7]
```diff
@@ -1,71 +1,35 @@
...
-{{#*inline "modal-content"}}
-  <div class="modal-content {{current}}-state-view fade{{#ifequal state current}} in{{/ifequal}}">
-    {{> @partial-block }}
-  </div>
-{{/inline}}
-
 <div class="modal-dialog">
-  {{#> modal-content current='default'}}
-    {{> header default}}
+  {{#> progress-dialog/modal-content current='default'}}
```
Comment (eugenet8k) [inline src/templates/progress-dialog.hbs]: My previous comment was about putting all the templates related to progress dialog in one folder `/templates/progres-dialog/' therefore we could avoid having excessive usage of `progress-dialog/...` all other this template

File: src/lib/view-helper.coffee (hunk 6 above)
Comment (bzalasky) [inline src/lib/view-helper.coffee]: Looks like I missed this file when I switched back to CommonJS.

File: src/lib/mixin-builder.coffee (hunk 5 above)
Comment (bzalasky) [inline src/lib/mixin-builder.coffee]: You can do it either way. I've seen both. The form you're using here is common when you're returning a higher-order component in React. Simplest way to think about HoCs is that they're components wrapping another component with additional functionality, similar to how we use mixins. That wrapping would happen after the class has been defined.

File: src/templates/progress-dialog.hbs (hunk 7 above)
Comment (bzalasky) [inline src/templates/progress-dialog.hbs]: Oh, I'll make that change.... how do you want to handle progress-dialog.hbs? progress-dialog/progress-dialog.hbs is kind of obnoxious. But cleaning up the actual template seems more important than worrying about that path. Perhaps progress-dialog/container.hbs?

Comment (eugenet8k) [inline src/templates/progress-dialog.hbs]: `progress-dialog/progress-dialog.hbs` is okay I think. Its better this way for fuzzy searches etc.

File: README.md [hunk 8]
```diff
@@ -1,10 +1,13 @@
 # Gringotts [![Build Status](https://travis-ci.org/lookout/gringotts.png?branch=master)](https://travis-ci.org/lookout/gringotts)
 
-A collection of [real mixins](http://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/)
-and utilities supporting common behaviors found in web apps built with `Chaplin` and `Handlebars`.
+A collection
+of
+[real mixins](http://justinfagnani.com/2015/12/21/real-mixins-with-javascript-classes/) and
```
Comment (eugenet8k) [inline README.md]: What's the point of having line breaks here?

File: package.json (hunk 3 above)
Comment (bzalasky) [inline package.json]: `/bin` is the convention I've seen and used in the past, but I'm not attached to it.

File: README.md (hunk 8 above)
Comment (bzalasky) [inline README.md]: It makes it so you can actually read it in your editor. It's done by emacs via fill-paragraph, but there are similar options in other editors.

Comment (bzalasky) [inline README.md]: I agree the diff looks odd here though.

Comment (bzalasky) [inline README.md]: I'll double check the file to make sure nothing funny happened.

Comment (bzalasky) [inline README.md]: Yeah, it's just the long link messing it up. I'll switch to referenced links.

**Comment by eugenet8k** [comment]
Will there be a way, to just build the project? I see now the build command actually triggers unit tests run. I am not sure if we even need to consider building as we are going to distribute gringotts in coffeescript format. In that case it's not clear to me the idea behind moving files into `dist/` folder? Shouldn't we just publish npm package with root pointed to gringotts/src folder?

File: .gitignore [hunk 9]
```diff
@@ -1,8 +1,6 @@
 .DS_Store
-.grunt/
-docs/
+.DSStore
 node_modules/
-public/
-tmp/*
 npm-debug.log
-/test-results.xml
```
Comment (eugenet8k) [inline .gitignore]: From what I remember `test-results.xml` is being picked up by different build systems like (Jenkins) to read the unit tests passing stats and maintain this statistic over time. So we may consider keep this feature, therefore not excluding this line from ignores.

Comment (bzalasky) [reply .gitignore]: I don't believe this file is being generated by Karma.

File: .gitignore [hunk 10]
```diff
@@ -1,8 +1,6 @@
...
+coverage/
+dist/
```
Comment (eugenet8k) [inline .gitignore]: do we generate anything into `/dist` folder at the moment?

File: Gruntfile.coffee [hunk 11]
```diff
@@ -1,271 +0,0 @@
...
-          "cp -R -v public/src #{buildPath};"
-
-    connect:
-      server:
-        options:
-          base: ['public', 'test']
-          port: 8000
-          useAvailablePort: true
-
-    checkDependencies:
```
Comment (eugenet8k) [inline Gruntfile.coffee]: I wonder if there is a way to check that npm dependencies are update before we start any webpack building? This is the line that did the same in grunt builds.

File: lib/mixin-builder.coffee [hunk 12]
```diff
@@ -0,0 +1,19 @@
+mixinHelper = require './mixin-helper'
```
Comment (eugenet8k) [inline lib/mixin-builder.coffee]: I whish we could have two commits, one that moves all source files our of `/src` folder and another that doesn't actual webpack update. Now the history is pretty messed up, all files are being recreated, so git blaming will be hard in the long run.

File: package.json [hunk 13]
```diff
@@ -19,53 +20,62 @@
...
+    "lib/",
+    "mixins/",
+    "models/",
+    "templates/",
+    "views/"
+  ],
   "scripts": {
-    "start": "grunt",
-    "test": "grunt test-ci"
+    "start": "scripts/test.sh",
```
Comment (eugenet8k) [inline package.json]: that file seems to have only one line of command, could we move it here?

File: package.json [hunk 14]
```diff
@@ -19,53 +20,62 @@
...
+    "lint": "scripts/lint.sh",
+    "lint:coffee": "echo lib mixins models templates test views | xargs -n 1 coffeelint -f coffeelint.json",
+    "test": "scripts/test-ci.sh",
```
Comment (eugenet8k) [inline package.json]: same here

File: package.json [hunk 15]
```diff
@@ -19,53 +20,62 @@
...
-    "grunt-mocha-blanket": "0.6.8",
-    "grunt-newer": "1.2.0",
-    "grunt-notify": "0.4.5",
-    "grunt-shell": "2.1.0",
-    "grunt-string-replace": "1.3.1",
-    "grunt-verbosity": "1.0.1",
-    "grunt": "1.0.1",
-    "load-grunt-tasks": "3.5.2",
+    "handlebars-loader": "1.5.0",
+    "htmlhint-loader": "^1.3.0",
```
Comment (eugenet8k) [inline package.json]: I wouldn't mind killing "^" here

File: package.json (hunk 13 above)
Comment (bzalasky) [reply package.json]: Will do.

File: lib/mixin-builder.coffee (hunk 12 above)
Comment (bzalasky) [reply lib/mixin-builder.coffee]: I used git mv, so it should count them as renames... I don't think we'll lose history. Though I will say, Github's blame doesn't always match what you get from git in the command line.

File: .gitignore (hunk 9 above)
Comment (bzalasky) [reply .gitignore]: I'll add the JUnit reporter to gringotts as well, and add test-results.xml back to .gitignore.

File: .gitignore (hunk 10 above)
Comment (bzalasky) [reply .gitignore]: I think I can ditch it at this point. I went with a whitelist approach in package.json that accomplishes what dist was doing for us before with the bash script.

File: package.json (hunk 15 above)
Comment (bzalasky) [reply package.json]: Will do.

**Comment by eugenet8k** [comment]
@bzalasky it seems somehow the file `.nvmrc` still has old version of node `6.9.2`, where I guess we want it `8.4.0`.

File: package.json (hunk 13 above)
Comment (eugenet8k) [reply package.json]: Is it still in progress?

File: package.json [hunk 16]
```diff
@@ -19,53 +20,62 @@
...
+    "views/"
+  ],
   "scripts": {
-    "start": "grunt",
-    "test": "grunt test-ci"
+    "start": "scripts/test.sh",
+    "lint": "scripts/lint.sh",
+    "lint:coffee": "echo lib mixins models templates test views | xargs -n 1 coffeelint -f coffeelint.json",
+    "test": "scripts/test-ci.sh",
+    "build": "scripts/build.sh",
```
Comment (eugenet8k) [inline package.json]: I couldn't find this file in the repo.

Comment (bzalasky) [reply package.json]: I removed the file. files array in package.json does this for us on publish now. I'll remove this as well.

File: package.json (hunk 13 above)
Comment (bzalasky) [reply package.json]: Done.

File: package.json (hunk 14 above)
Comment (bzalasky) [reply package.json]: Done.

File: package.json (hunk 15 above)
Comment (bzalasky) [reply package.json]: Done.

File: package.json (hunk 16 above)
Comment (bzalasky) [reply package.json]: Done.

File: Gruntfile.coffee (hunk 11 above)
Comment (bzalasky) [reply Gruntfile.coffee]: I'll add this to my list of follow on tasks.

File: .gitignore (hunk 10 above)
Comment (bzalasky) [reply .gitignore]: Done.

**Pull request merged**